[Navigation API] Fix 16% PLT regression by removing Sync IPC

RupinMittal opened the pull request:
#### b539f239da5e0740c16d7ca1dd9dd9b0b29bf72a
<pre>
[Navigation API] Fix 16% PLT regression by removing Sync IPC
<a href="https://bugs.webkit.org/show_bug.cgi?id=298498">https://bugs.webkit.org/show_bug.cgi?id=298498</a>
<a href="https://rdar.apple.com/160000632">rdar://160000632</a>

Reviewed by NOBODY (OOPS!).

When a new Document is created, it&apos;s Navigation object&apos;s list of history entries
must be initialized. In the initial implementation of the Navigation API in
273532@main, this was done by making a sync IPC to the UI Process to get each
item in the b/f list. (The calls to itemAtIndex()). 298922@main changed this so
the whole list is retrieved in just one sync IPC call (The call to itemsForFrame()).

So enabling the Navigation API means that every time a new Document is created,
a new sync IPC is sent. This is terrible for performance. It&apos;s a ~16% PLT regression.

To fix this, this patch removes this extra sync IPC. The list is still needed, but
now it&apos;s retreived in an IPC that already occurs when a navigation results in a
document&apos;s creation: decidePolicyForNavigationAction.

When a load is triggered, the Web Process sends this IPC to the UI Process to
determine if the navigation should continue. We modify the returned PolicyDecision
object to also contain the b/f list. This list is be stored on the Frame and passed
into the Navigation object when we initialize it for the new Document.

There are some nuances that need fixing to make this work:

1. Missing the navigation-triggering HistoryItem.

   When a navigation occurs, the Web Process only creates and sends the HistoryItem
   associated with this navigation to the UI Process once the UI Process has told
   the Web Process to continue the navigation via decidePolicyForNavigationAction.
   This means that at the time where the b/f list is retrieved via IPC, it will be
   missing the HistoryItem that triggered this navigation. The list is incomplete.

   To fix this, the Web Process appends this missing item to the list once it has
   created it. This occurs in two places:

       1. HistoryController::updateBackForwardListClippedAtTarget()

              Some situations where this is called:
              - The initial load of a main frame
              - A navigation (in either a main frame or subframe) via
                navigation.navigate().

       2. HistoryController::updateForRedirectWithLockedBackForwardList()

              Some situations where this is called:
              - If a main frame navigates, the new HistoryItems for its
                subframes will be created here.
              - Scenarios where a load is of type RedirectWithLockedBackForwardList,
                such as a frame loading but then redirecting to another item
                immediately and without user interaction. In these situations, the
                new HistoryItem replaces the old one, so we replace the item in the
                list rather than append.

              * On a iframe&apos;s load, the HistoryItem created for it here will have
                the same ItemID as the currentItem of it&apos;s parent frame&apos;s
                HistoryController object. This may be stale if the iframe is
                navigating but the parent frame is not. In that case, the HistoryItem
                created for the iframe will be wrong. This scenario occurs in the test
                anchor-fragment-form-submit-withpath.html. To fix this, we ensure that
                when a frame navigates, the currentItems for the other frames in the
                frame tree are also updated.

2. Clones

   Consider two sibling frames A and B. If A navigates, a new history tree
   is created, which will have a HistoryItem for B as well. This item will be
   a clone of the HistoryItem for B in the first history tree. When the list
   of all HistoryItems for B is obtained, it will contain both HistoryItems.
   But B cannot go back and forward between the two items. So we must remove
   the clone.

3. Coalescing with the list from the previous Document&apos;s Navigation object.

   When initializing the Navigation object for a non-main frame, the list is
   initially holds the previous Document&apos;s entries. We must coalesce this list
   with the list received from the UI Process.

   This should have been happening before, but it turns out that prior to this
   patch, due to incorrect logic in itemsForFrame, the list received via sync
   IPC contained no entries for non-main frames.So many tests were passing by
   mistake. In some cases, tests were passing, but the underlying state of
   m_entries was wrong (contained duplicate items).

   The logic to coalesce the list from the previous Document with the list from
   IPC is:

       For all items in the IPC list:
           If it&apos;s already in the previous list, don&apos;t add it **
           Otherwise, append it.

       ** Unless the item in the IPC list has a different URL. This represents the
          scenario where an item has been updated with a locked back forward list
          as described above. Then the new item would have replaced the last item.
          The UI Process will know about this but the previous Document&apos;s Navigation
          object (which is in the Web Process) may not. So in this case, we replace
          the previous entry with the new one from IPC.

The overarching idea is that this list must match the UI Process b/f list. The
UI Process b/f list is the source of truth. If we were to sync IPC to the UI
Process to get the list, it would be correct. But we cannot due to performance.
And since the UI Process&apos;s list is not fully updated at the time of the
decidePolicyForNavigationAction IPC, the Web Process must do the above steps
to reconstruct the missing/updated items to reflect the state the UI Process
list will soon be in.

Bugs found while doing this:
1. navigation.navigate(&lt;url&gt;, { history: &apos;push&apos; }) doesn&apos;t always create a new
   HistoryItem like it should. This means we may be short an item. This causes
   a test to fail. It was passing accidently before.
2. Navigations that cause forward pruning in iframes may not prune the UI Process
   b/f list, which means the list retreived via IPC in the next navigation will
   be incorrect. This causes another test to fail. It was passing accidently before.
3. navigation.navigate(&lt;url&gt;, { history: &apos;replace&apos; }) doesn&apos;t always replace
   the current HistoryItem in the UI Process list, it may instead create a new
   HistoryItem. This is not causing any tests to fail for now, but must be fixed.

* LayoutTests/TestExpectations:

Mark the two tests as failing as described above. They were passing accidently
before this patch.

* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::allItemsForFrame):
(WebCore::BackForwardController::allItems): Deleted.
(WebCore::BackForwardController::itemsForFrame): Deleted.
(WebCore::BackForwardController::reachableItemsForFrame): Deleted.
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::setItemID):
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument):

When initializing the Navigation object for the new Document,
pass in the b/f list obtained via IPC here.

* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::updateForRedirectWithLockedBackForwardList):

This is one location where a new HistoryItem is created after the b/f list
has already been retreived via IPC. So add the missing item.

(WebCore::HistoryController::updateForCommit):

Ensure that on a navigation, the current item for all frames is updated
to have the correct itemID.

(WebCore::HistoryController::recursiveSetCurrentItemID):
(WebCore::HistoryController::updateBackForwardListClippedAtTarget):

This is one location where a new HistoryItem is created after the b/f list
has already been retreived via IPC. So add the missing item.

* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledHistoryNavigationByKey::findBackForwardItemByKey):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::backForwardList const):
(WebCore::LocalFrame::setBackForwardList):
(WebCore::LocalFrame::addItemToBackForwardList):

Store the b/f list here. One alternative considered was to store it
on the Navigation object itself. But that object is owned by the
Document and the Document may not exist at the time of the
decidePolicyForNavigationAction IPC.

* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::getEntryIndexOfHistoryItem):
(WebCore::Navigation::initializeForNewWindow):

Add new logic for removing clones from the IPC list,
and coalescing with the list from the previous Document.

* Source/WebCore/page/Navigation.h:
* Source/WebKit/Shared/PolicyDecision.h:

The PolicyDecision will now also contain the b/f list.

* Source/WebKit/Shared/PolicyDecision.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationActionAsync):
(WebKit::WebPageProxy::decidePolicyForNavigationActionSync):
(WebKit::WebPageProxy::backForwardAllItemsForFrame):
(WebKit::WebPageProxy::backForwardAllItems): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):

Obtain the b/f list from the UI Process in this IPC.

* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::allItemsForFrame):
(WebKit::WebBackForwardListProxy::allItems): Deleted.
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::setBackForwardList):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::allItemsForFrame):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b539f239da5e0740c16d7ca1dd9dd9b0b29bf72a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119773 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39466 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30117 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126074 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71845 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d26ad8c7-5e67-4362-aed4-bd6b13ffac74) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121649 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40161 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48043 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90970 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60251 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/23325547-ca3e-48fc-bcaf-5a118272d239) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122725 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32061 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107399 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71525 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c77ffa97-b646-4ee6-ab60-bd88d32b1355) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31091 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25502 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69718 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101532 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25694 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129019 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46693 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35379 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99568 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47059 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103590 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99412 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44853 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22867 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43265 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46555 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52261 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46021 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49370 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47707 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7844,6 +7844,18 @@ void WebPageProxy::beginSafeBrowsingCheck(const URL&, API::Navigation&, bool for
 }
 #endif
 
+void WebPageProxy::decidePolicyForNavigationActionAndGetBackForwardListAsync(IPC::Connection& connection, NavigationActionData&& data, FrameIdentifier frameID, CompletionHandler<void(PolicyDecision&&, Vector<Ref<FrameState>>&&)>&& completionHandler)
```
Seems like this should just outright replace `decidePolicyForNavigationActionAsync` but I presume you separated this out to make it easier to iterate on changes?

bnham commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1599,6 +1599,21 @@ RefPtr<SecurityOrigin> LocalFrame::frameDocumentSecurityOrigin() const
+
+void LocalFrame::addItemToBackForwardList(Ref<HistoryItem>&& item)
+{
+    m_backForwardList.append(WTFMove(item));
```
Is this correct? I think there are some cases where you need to add an item to the middle of a back-forward list and then toss everything from that point in the list moving forward. (See `WebBackForwardList::addItem`.)

bnham reviewed (commented):
Seems like the same-origin filtering logic in `Navigation::initializeForNewWindow` should be moved to the UIProcess side for this patch instead, but that can be done later. Right now it's best to prioritize whether this logic works correctly or not.

charliewolfe reviewed (commented):
If this solution is needed as a temporary fix to enable the Navigation API, fine. But it’s pretty ugly and a step in the wrong direction. I think the correct way to fix this is to move more of the Navigation API logic into the UI process. It was a mistake to implement it in a way that requires the web process to have full knowledge of the back/forward list.

RupinMittal committed:
[Navigation API] Fix 16% PLT regression by removing Sync IPC
https://bugs.webkit.org/show_bug.cgi?id=298498
rdar://160000632

Reviewed by NOBODY (OOPS!).

When a new Document is created, it's Navigation object's list of history entries
must be initialized. In the initial implementation of the Navigation API in
273532@main, this was done by making a sync IPC to the UI Process to get each
item in the b/f list. (The calls to itemAtIndex()). 298922@main changed this so
the whole list is retrieved in just one sync IPC call (The call to itemsForFrame()).

So enabling the Navigation API means that every time a new Document is created,
a new sync IPC is sent. This is terrible for performance. It's a ~16% PLT regression.

To fix this, this patch removes this extra sync IPC. The list is still needed, but
now it's retreived in an IPC that already occurs when a navigation results in a
document's creation: decidePolicyForNavigationAction.

When a load is triggered, the Web Process sends this IPC to the UI Process to
determine if the navigation should continue. We modify the returned PolicyDecision
object to also contain the b/f list. This list is be stored on the Frame and passed
into the Navigation object when we initialize it for the new Document.

There are some nuances that need fixing to make this work:

1. Missing the navigation-triggering HistoryItem.

   When a navigation occurs, the Web Process only creates and sends the HistoryItem
   associated with this navigation to the UI Process once the UI Process has told
   the Web Process to continue the navigation via decidePolicyForNavigationAction.
   This means that at the time where the b/f list is retrieved via IPC, it will be
   missing the HistoryItem that triggered this navigation. The list is incomplete.

   To fix this, the Web Process appends this missing item to the list once it has
   created it. This occurs in two places:

       1. HistoryController::updateBackForwardListClippedAtTarget()

              Some situations where this is called:
              - The initial load of a main frame
              - A navigation (in either a main frame or subframe) via
                navigation.navigate().

       2. HistoryController::updateForRedirectWithLockedBackForwardList()

              Some situations where this is called:
              - If a main frame navigates, the new HistoryItems for its
                subframes will be created here.
              - Scenarios where a load is of type RedirectWithLockedBackForwardList,
                such as a frame loading but then redirecting to another item
                immediately and without user interaction. In these situations, the
                new HistoryItem replaces the old one, so we replace the item in the
                list rather than append.

              * On a iframe's load, the HistoryItem created for it here will have
                the same ItemID as the currentItem of it's parent frame's
                HistoryController object. This may be stale if the iframe is
                navigating but the parent frame is not. In that case, the HistoryItem
                created for the iframe will be wrong. This scenario occurs in the test
                anchor-fragment-form-submit-withpath.html. To fix this, we ensure that
                when a frame navigates, the currentItems for the other frames in the
                frame tree are also updated.

2. Clones

   Consider two sibling frames A and B. If A navigates, a new history tree
   is created, which will have a HistoryItem for B as well. This item will be
   a clone of the HistoryItem for B in the first history tree. When the list
   of all HistoryItems for B is obtained, it will contain both HistoryItems.
   But B cannot go back and forward between the two items. So we must remove
   the clone.

3. Coalescing with the list from the previous Document's Navigation object.

   When initializing the Navigation object for a non-main frame, the list is
   initially holds the previous Document's entries. We must coalesce this list
   with the list received from the UI Process.

   This should have been happening before, but it turns out that prior to this
   patch, due to incorrect logic in itemsForFrame, the list received via sync
   IPC contained no entries for non-main frames.So many tests were passing by
   mistake. In some cases, tests were passing, but the underlying state of
   m_entries was wrong (contained duplicate items).

   The logic to coalesce the list from the previous Document with the list from
   IPC is:

       For all items in the IPC list:
           If it's already in the previous list, don't add it **
           Otherwise, append it.

       ** Unless the item in the IPC list has a different URL. This represents the
          scenario where an item has been updated with a locked back forward list
          as described above. Then the new item would have replaced the last item.
          The UI Process will know about this but the previous Document's Navigation
          object (which is in the Web Process) may not. So in this case, we replace
          the previous entry with the new one from IPC.

The overarching idea is that this list must match the UI Process b/f list. The
UI Process b/f list is the source of truth. If we were to sync IPC to the UI
Process to get the list, it would be correct. But we cannot due to performance.
And since the UI Process's list is not fully updated at the time of the
decidePolicyForNavigationAction IPC, the Web Process must do the above steps
to reconstruct the missing/updated items to reflect the state the UI Process
list will soon be in.

Bugs found while doing this:
1. navigation.navigate(<url>, { history: 'push' }) doesn't always create a new
   HistoryItem like it should. This means we may be short an item. This causes
   a test to fail. It was passing accidently before.
2. Navigations that cause forward pruning in iframes may not prune the UI Process
   b/f list, which means the list retreived via IPC in the next navigation will
   be incorrect. This causes another test to fail. It was passing accidently before.
3. navigation.navigate(<url>, { history: 'replace' }) doesn't always replace
   the current HistoryItem in the UI Process list, it may instead create a new
   HistoryItem. This is not causing any tests to fail for now, but must be fixed.

* LayoutTests/TestExpectations:

Mark the two tests as failing as described above. They were passing accidently
before this patch.

* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::allItemsForFrame):
(WebCore::BackForwardController::allItems): Deleted.
(WebCore::BackForwardController::itemsForFrame): Deleted.
(WebCore::BackForwardController::reachableItemsForFrame): Deleted.
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::setItemID):
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument):

When initializing the Navigation object for the new Document,
pass in the b/f list obtained via IPC here.

* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::updateForRedirectWithLockedBackForwardList):

This is one location where a new HistoryItem is created after the b/f list
has already been retreived via IPC. So add the missing item.

(WebCore::HistoryController::updateForCommit):

Ensure that on a navigation, the current item for all frames is updated
to have the correct itemID.

(WebCore::HistoryController::recursiveSetCurrentItemID):
(WebCore::HistoryController::updateBackForwardListClippedAtTarget):

This is one location where a new HistoryItem is created after the b/f list
has already been retreived via IPC. So add the missing item.

* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledHistoryNavigationByKey::findBackForwardItemByKey):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::backForwardList const):
(WebCore::LocalFrame::setBackForwardList):
(WebCore::LocalFrame::addItemToBackForwardList):

Store the b/f list here. One alternative considered was to store it
on the Navigation object itself. But that object is owned by the
Document and the Document may not exist at the time of the
decidePolicyForNavigationAction IPC.

* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::getEntryIndexOfHistoryItem):
(WebCore::Navigation::initializeForNewWindow):

Add new logic for removing clones from the IPC list,
and coalescing with the list from the previous Document.

* Source/WebCore/page/Navigation.h:
* Source/WebKit/Shared/PolicyDecision.h:

The PolicyDecision will now also contain the b/f list.

* Source/WebKit/Shared/PolicyDecision.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationActionAsync):
(WebKit::WebPageProxy::decidePolicyForNavigationActionSync):
(WebKit::WebPageProxy::backForwardAllItemsForFrame):
(WebKit::WebPageProxy::backForwardAllItems): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):

Obtain the b/f list from the UI Process in this IPC.

* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::allItemsForFrame):
(WebKit::WebBackForwardListProxy::allItems): Deleted.
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::setBackForwardList):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::allItemsForFrame):

basuke commented on Source/WebCore/history/BackForwardController.cpp:
```diff
@@ -175,32 +175,9 @@ RefPtr<HistoryItem> BackForwardController::itemAtIndex(int i, std::optional<Fram
-{
-    // Returns only the frame items that correspond to the currently reachable session history.
-    // This is different from itemsForFrame() which returns all frame items across the frame's lifetime.
-    Vector<Ref<HistoryItem>> reachableFrameItems;
```
How do you recover this concept of reachability by deleting this method? I cannot find the alternative way.

RupinMittal replied:
Despite what the comment says, `reachableItemsForFrame` was indeed getting all the items for that frame across its lifetime. So that comment is wrong.

`allItemsForFrame(FrameIdentifier frameID)` already does this same thing. So there is no change in behavior if we use `allItemsForFrame` in `CachedPage` instead of `reachableItemsForFrame`.

RupinMittal replied:
Although actually what I think we should do here is filter based on the `currentItem`'s origin the same way we do in `initializeForWindow`. Otherwise we may end with a list that contains extra items of different origins that should not be in the list. I'll look into this.

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1599,6 +1599,28 @@ RefPtr<SecurityOrigin> LocalFrame::frameDocumentSecurityOrigin() const
     return nullptr;
 }
 
+Vector<Ref<HistoryItem>> LocalFrame::backForwardList() const
```
Should return a `const Vector<Ref<HistoryItem>>&`

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1599,6 +1599,28 @@ RefPtr<SecurityOrigin> LocalFrame::frameDocumentSecurityOrigin() const
+void LocalFrame::addItemToBackForwardList(Ref<HistoryItem>&& item)
+{
+    for (size_t i = 0; i < m_backForwardList.size(); i++) {
+        if (m_backForwardList[i]->itemID().object() <=> item->itemID().object() == std::strong_ordering::equal) {
```
this is normally written as:
```
is_eq(m_backForwardList[i]->itemID().object(), item->itemID().object())
```

cdumez commented on Source/WebCore/page/LocalFrame.h:
```diff
@@ -376,6 +381,8 @@ class LocalFrame final : public Frame {
 
     UniqueRef<ScriptController> m_script;
 
+    Vector<Ref<HistoryItem>> m_backForwardList;
```
Hmm, I did not look much into this yet but this looks a little fishy that the LocalFrame now has a copy of the history items. This is supposed to live in the HistoryController.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -111,7 +111,7 @@ static std::optional<size_t> getEntryIndexOfHistoryItem(const Vector<Ref<Navigat
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#initialize-the-navigation-api-entries-for-a-new-document
-void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType> navigationType, LocalDOMWindow* previousWindow)
+void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType> navigationType, LocalDOMWindow* previousWindow, Vector<Ref<HistoryItem>> backForwardList)
```
We should not be passing the Vector by value here.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -169,7 +169,17 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
-    auto rawEntries = page->backForward().itemsForFrame(frame()->frameID());
+    Vector<Ref<HistoryItem>> rawEntries;
+    for (auto entry : backForwardList) {
+        if (!rawEntries.size()) {
```
`rawEntries.isEmpty()`

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -197,12 +207,29 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
 
-    size_t start = m_entries.size();
+    if (m_entries.size() > 0) {
+        auto last = m_entries[m_entries.size() - 1];
```
`auto last = m_entries.last();`

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -197,12 +207,29 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
+            // If the frame was redirected with the b/f history locked, the redirected
+            // item replaces the last item in the b/f list. To keep m_entries in sync
+            // with the UI process b/f list, we must do the same replacement here.
+            if ((item->itemID().object() <=> lastItemID.object()) == std::strong_ordering::equal && item->url() != last->associatedHistoryItem().url()) {
```
Use `is_eq()`

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -197,12 +207,29 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
 
-    for (Ref item : items)
-        m_entries.append(NavigationHistoryEntry::create(*this, WTFMove(item)));
+            if ((item->itemID().object() <=> lastItemID.object()) == std::strong_ordering::greater)
```
use `is_gt()`

cdumez reviewed (commented):
I did not have time to finish reviewing this.
